dataservice/api/common/pagination.py

```diff
@@ -101,25 +102,41 @@ def indexd_pagination(q, after, limit):
 
     :returns: A Pagination object
     """
+    def prefetch_indexd(after):
+        """ Compute dids for the page and have indexd fetch them in bulk """
+        model = q._entities[0].mapper.entity
+        gfs = (q.order_by(model.created_at.asc())
+                .filter(model.created_at > after)
+                .with_entities(model.latest_did)
+                .limit(limit).all())
+        dids = [gf[0] for gf in gfs]
+        indexd.prefetch(dids)
+
+    indexd.clear_cache()
+    prefetch_indexd(after)
     pager = Pagination(q, after, limit)
     keep = []
     refresh = True
     next_after = None
     # Continue updating the page until we get a page with no deleted files
     while (pager.total > 0 and refresh):
         refresh = False
-        # Move the cursor ahead to the last valid file
-        next_after = keep[-1].created_at if len(keep) > 0 else after
-        # Number of results needed to fulfill the original limit
-        remain = limit - len(keep)
-        pager = Pagination(q, next_after, remain)
-
         for st in pager.items:
             if hasattr(st, 'was_deleted') and st.was_deleted:
```

Comment:
Isn't this never going to happen now? Before the loop, you first populate the indexd page cache. Then you construct the Pagination obj, which results in a bunch of calls to indexd.get() which results in a bunch of indexd page cache lookups which means the requests to indexd never go out. So then you won't know if a file was deleted?

Reply:
I believe this will still happen because the `merge_indexd()` is still called for every object in the page, only, it reads from the cache if available. If it's not in the cache, it will attempt to get it from indexd, in which case, it will return 404 and mark it as `was_deleted`.
Will look more closely at this to confirm.